metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -354,19 +387,113 @@ final class FileDecoderProvider(
+      pathInfo: PathInfo
+  ): Option[PresentationCompiler] =
     for {
-      pc <- Future {
-        for {
-          targetId <- pathInfo.targetId
-          pc <- compilers.loadCompiler(targetId)
-        } yield pc
+      targetId <- pathInfo.targetId
+      pc <- compilers.loadCompiler(targetId)
+    } yield pc
+
+  /**
+   * For clients supporting executing commands [[TastyResponse]] is returned and clients can determine on their own how to handle returned value.
+   * If client supports http, he is redirected to the tasty endpoint defined at [[MetalsHttpServer]].
+   * That endpoint reuses logic declared in [[TastyHandler]]
+   * In both cases logic is pretty same:
+   * - for a given URI (which could be .scala or .tasty file itself) try to find .tasty file
+   * - dispatch request to the Presentation Compiler. It's worth noting that PC takes into account
+   *   client configuration to determine proper response format (HTML, console or plain text)
+   */
+  def executeShowTastyCommand(
+      params: l.TextDocumentPositionParams
+  ): Future[Unit] = {
+    val uri = new URI(params.getTextDocument().getUri())
+    val position = params.getPosition()
+    if (
+      clientConfig.isExecuteClientCommandProvider() && !clientConfig
+        .isHttpEnabled()
+    ) {
+      val response = getTastyForURI(uri, Some(position)).map { result =>
+        DecoderResponse(
+          uri.toString(),
+          result.fold(_ => null, identity),
+          result.fold(identity, _ => null)
+        )
       }
-      output <- pc match {
-        case None => Future.successful(None)
-        case Some(pc) =>
-          pc.getTasty(pathInfo.path.toURI, false, false).asScala.map(Some(_))
+      response.map { tasty =>
+        val command = new l.ExecuteCommandParams(
+          "metals-show-tasty",
+          List[Object](tasty).asJava
+        )
+        languageClient.metalsExecuteClientCommand(command)
+        scribe.debug(s"Executing show TASTy $command")
+      }
+    } else
+      httpServer() match {
+        case Some(server) =>
+          Future.successful(
+            Urls.openBrowser(server.address + s"/tasty?file=$uri")
+          )
+        case None =>
+          Future.successful {
+            scribe.warn(
+              "Unable to run show tasty. Make sure `isHttpEnabled` is set to `true`."
+            )
+          }
       }
-    } yield output
+  }
+
+  def getTastyForURI(
+      uri: URI,
+      cursorPosition: Option[l.Position] = None
+  ): Future[Either[String, String]] = {
+    val error = s"Can't find existing build target for $uri"
+    val pcAndTargetUri =
+      for {
+        path <- Try(AbsolutePath.fromAbsoluteUri(uri)) match {
+          case Success(path) if !path.isFile => Left(s"$uri doesn't exist")
+          case Success(path) if path.isFile => Right(path)
+          case Failure(_) => Left(s"$uri has to be absolute")
+        }
+        pathInfo <-
+          if (path.isScala)
+            cursorPosition
```

Comment:
So for `cursorPosition = None` we show nothing? Could we maybe use a default value for now that would work if a file is named the same as the tasty file? We could have a separate method that just finds a single or main class inside a file.

Reply:
There is already `.orElse(findClassesDirFileFromSource(path, "tasty"))` which do what you want ;)
Current PR is just a small improvement in the terms of cases when a position is supplied. Otherwise, the behavior is the same as before.

Reply:
Ach, right! Thanks!